tests/cmd/upgrade/preserve_precision_patch.toml

```diff
@@ -0,0 +1,13 @@
+bin.name = "cargo-upgrade"
+args = ["upgrade", "--preserve-precision"]
+status = "success"
+stdout = """
+cargo-list-test-fixture:
+"""
+stderr = """
+    Upgrading docopt v0.8.0 -> v99999.0.0
```

Comment:
Hmm, this is an interesting case.  Should precision be relative to the breaking field (minor in this case) or be absolute?

Reply:
Sorry, could you further explain what you mean by "breaking field?"

Reply:
In cargo's variant of semver,
- `x.0.0`: changes to major indicate breaking versions
- `0.x.0`: changes to minor indicate breaking versions
- `0.0.x`: changes to patch indicate breaking versions

Reply:
But why would minor be the "breaking field" in this case? (Sorry, if I am being dense.)

Reply:
The semver spec says that *any* version change for `0.x.y` cannot be assumed to be backwards compatible ([Item 4](https://semver.org/#spec-item-4)).

Cargo deviates from semver and treats the left-most non-zero field like major where if it is changed, then you assume its backwards incompatible but if any field to its right is changed then you assume its compatible.

When cargo resolves a `Cargo.toml` file, if two "backwards compatible" version requirements exist (`0.3.4` and `0.3.6`), it will choose a version that satisfies them all (anything that is `>=0.3.6,<0.4.0` in this case).  If a common version doesn't exist, it will error.  If two "backwards incompatible" version requirements exist (`0.3.4`, `0.4.2`), then cargo will build two different versions of that crate and make them available (this happens either when a dependency uses one or if you use the `package` field to rename one).  This allows your program to work even if incompatible versions are used in the dependency tree (e.g. clap3 has a transitive dependency on clap2 through criterion).

So major, minor, and patch correspond to the absolute position of the fields but which field is breaking depends on which fields are zero.

Reply:
OK, I understand. Thank you for the detailed explanation.

IMHO, the precision should be absolute.

Conceivably, either could be a user's preference/intent. Moreover, the choice could depend upon the dependency. So, for the sake of argument, let's assume that either possibility is equally likely.

Consider these two alternatives:
* A user wants relative precision, but `cargo-upgrade` uses absolute precision. Then `cargo-upgrade` generates unwanted noise for the user (e.g., more changes to their Cargo.toml file) until the user removes the unwanted, extra precision.
* A user wants absolute precision, but `cargo-upgrade` uses relative precision. Then `cargo-upgrade` generates no updates to the associated dependency until the user adds the desired, missing precision.

Given the choice between "an annoyance that can be turned off" and "silently failing," I think the former is the lesser evil.